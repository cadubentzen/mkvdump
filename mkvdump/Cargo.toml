[package]
name = "mkvdump"
version = "0.5.0"
authors = ["Carlos Bentzen <cadubentzen@gmail.com>"]
edition = "2021"
license = "MIT OR Apache-2.0"
description = "MKV and WebM parser CLI tool"
repository = "https://github.com/cadubentzen/mkvdump"
keywords = ["mkv", "matroska", "webm", "cli", "ebml"]
categories = ["command-line-utilities", "multimedia", "parser-implementations"]
exclude = ["tests/*", ".github/*"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
mkvparser = { path = "../mkvparser", version = "0.4.4" }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
serde_yaml = "0.9"
clap = { version = "4", features = ["derive"] }

[dev-dependencies]
assert_cmd = "2"
insta = { version = "1.18.1", features = ["yaml"] }
