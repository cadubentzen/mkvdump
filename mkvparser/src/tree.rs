///! The tree module contains helpers for building tree
///! structures from parsed elements
use serde::Serialize;

use crate::{Body, Element, Header, Id};

/// A Master Element that owns its children for diplaying
/// it in an element tree
#[derive(Debug, PartialEq, Serialize)]
pub struct MasterElement {
    #[serde(flatten)]
    header: Header,
    children: Vec<ElementTree>,
}

/// An Element Tree can either be a leaf or a Master
/// element.
#[derive(Debug, PartialEq, Serialize)]
#[serde(untagged)]
pub enum ElementTree {
    /// A Normal Element that represents a leaf in the tree
    Normal(Element),
    /// A Master Element contains more elements.
    Master(MasterElement),
}

impl Id {
    fn can_be_children_of(&self, other: &Id) -> bool {
        !matches!((self, other), (Id::Cluster, Id::Cluster) | (Id::Ebml, _))
    }
}

/// Build element trees from a series of elements
pub fn build_element_trees(elements: &[Element]) -> Vec<ElementTree> {
    let mut trees = Vec::<ElementTree>::new();

    let mut index = 0;
    while index < elements.len() {
        let element = &elements[index];
        match element.body {
            Body::Master => {
                // parse_header() already handles Unknown sizes.
                let mut size_remaining = element.header.body_size.unwrap_or(usize::MAX);

                let mut children = Vec::<Element>::new();
                while size_remaining > 0 {
                    index += 1;

                    if let Some(next_child) = elements.get(index) {
                        if !next_child.header.id.can_be_children_of(&element.header.id) {
                            index -= 1;
                            break;
                        }

                        size_remaining -= if let Body::Master = next_child.body {
                            // Master elements' body size should not count in the recursion
                            // as the children would duplicate the size count, so
                            // we only consider the header size on the calculation.
                            next_child.header.header_size
                        } else {
                            next_child
                                .header
                                .size
                                .expect("Only Master elements can have unknown size")
                        };
                        children.push(next_child.clone());
                    } else {
                        // Elements have ended before reaching the size of the master element
                        break;
                    }
                }
                trees.push(ElementTree::Master(MasterElement {
                    header: element.header.clone(),
                    children: build_element_trees(&children),
                }));
            }
            _ => {
                trees.push(ElementTree::Normal(element.clone()));
            }
        }
        index += 1;
    }
    trees
}

#[cfg(test)]
mod tests {
    use crate::Unsigned;

    use super::*;

    #[test]
    fn test_build_element_trees() {
        let elements = [
            Element {
                header: Header::new(Id::Ebml, 5, 31),
                body: Body::Master,
            },
            Element {
                header: Header::new(Id::EbmlVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(1)),
            },
            Element {
                header: Header::new(Id::EbmlReadVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(1)),
            },
            Element {
                header: Header::new(Id::EbmlMaxIdLength, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(4)),
            },
            Element {
                header: Header::new(Id::EbmlMaxSizeLength, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(8)),
            },
            Element {
                header: Header::new(Id::DocType, 3, 4),
                body: Body::String("webm".to_string()),
            },
            Element {
                header: Header::new(Id::DocTypeVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(4)),
            },
            Element {
                header: Header::new(Id::DocTypeReadVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(2)),
            },
        ];

        let expected = vec![ElementTree::Master(MasterElement {
            header: Header::new(Id::Ebml, 5, 31),
            children: vec![
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(1)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlReadVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(1)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlMaxIdLength, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(4)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlMaxSizeLength, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(8)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocType, 3, 4),
                    body: Body::String("webm".to_string()),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocTypeVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(4)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocTypeReadVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(2)),
                }),
            ],
        })];

        assert_eq!(build_element_trees(&elements), expected);
    }
}
