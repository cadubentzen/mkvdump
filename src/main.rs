#![doc = include_str!("../README.md")]

use clap::{Parser, ValueEnum};
use mkvdump::parse_elements_from_file;
use mkvparser::tree::build_element_trees;
use serde::Serialize;

#[doc(hidden)]
#[derive(Parser)]
#[clap(author, version, about, long_about = None)]
struct Args {
    /// Name of the MKV/WebM file to be parsed
    filename: String,

    /// Output format
    #[clap(value_enum, short, long, default_value = "yaml")]
    format: Format,

    /// Add element positions in the output
    #[clap(short = 'p', long)]
    show_element_positions: bool,

    /// Show output as a sequence, rather than a tree
    #[clap(short = 'l', long)]
    linear_output: bool,
}

#[doc(hidden)]
#[derive(ValueEnum, Clone, PartialEq, Eq)]
enum Format {
    Json,
    Yaml,
}

#[doc(hidden)]
fn print_serialized<T: Serialize>(elements: &[T], format: &Format) {
    let serialized = match format {
        Format::Json => serde_json::to_string_pretty(elements).unwrap(),
        Format::Yaml => serde_yaml::to_string(elements).unwrap(),
    };
    println!("{}", serialized);
}

#[doc(hidden)]
fn main() -> anyhow::Result<()> {
    let args = Args::parse();
    let elements = parse_elements_from_file(&args.filename, args.show_element_positions)?;

    if args.linear_output {
        print_serialized(&elements, &args.format);
    } else {
        let element_trees = build_element_trees(&elements);
        print_serialized(&element_trees, &args.format);
    }

    Ok(())
}
