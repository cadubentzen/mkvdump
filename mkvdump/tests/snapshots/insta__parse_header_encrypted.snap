---
source: src/main.rs
assertion_line: 894
expression: "parse_buffer_to_end(include_bytes!(\"../inputs/encrypted.hdr\"))"
---
- id: EBML
  header_size: 5
  size: 36
  children:
    - id: EBMLVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLReadVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLMaxIDLength
      header_size: 3
      size: 4
      value: 4
    - id: EBMLMaxSizeLength
      header_size: 3
      size: 4
      value: 8
    - id: DocType
      header_size: 3
      size: 7
      value: webm
    - id: DocTypeVersion
      header_size: 3
      size: 4
      value: 4
    - id: DocTypeReadVersion
      header_size: 3
      size: 4
      value: 2
- id: Segment
  header_size: 12
  size: Unknown
  children:
    - id: Void
      header_size: 2
      size: 89
      value: ~
    - id: Info
      header_size: 5
      size: 50
      children:
        - id: TimestampScale
          header_size: 4
          size: 7
          value: 1000000
        - id: MuxingApp
          header_size: 3
          size: 19
          value: Bitmovin Encoder
        - id: WritingApp
          header_size: 3
          size: 19
          value: Bitmovin Encoder
    - id: Tracks
      header_size: 5
      size: 108
      children:
        - id: TrackEntry
          header_size: 2
          size: 103
          children:
            - id: TrackNumber
              header_size: 2
              size: 3
              value: 1
            - id: TrackUID
              header_size: 3
              size: 4
              value: 1
            - id: TrackType
              header_size: 2
              size: 3
              value: video
            - id: CodecID
              header_size: 2
              size: 7
              value: V_VP9
            - id: CodecPrivate
              header_size: 3
              size: 15
              value: "[01 01 00 02 01 14 03 01 08 04 01 01]"
            - id: ContentEncodings
              header_size: 3
              size: 51
              children:
                - id: ContentEncoding
                  header_size: 3
                  size: 48
                  children:
                    - id: ContentEncodingOrder
                      header_size: 3
                      size: 4
                      value: 0
                    - id: ContentEncodingScope
                      header_size: 3
                      size: 4
                      value: Block
                    - id: ContentEncodingType
                      header_size: 3
                      size: 4
                      value: Encryption
                    - id: ContentEncryption
                      header_size: 3
                      size: 33
                      children:
                        - id: ContentEncAlgo
                          header_size: 3
                          size: 4
                          value: AES
                        - id: ContentEncKeyID
                          header_size: 3
                          size: 19
                          value: "[eb 67 6a bb cb 34 5e 96 bb cf 61 66 30 f1 a3 da]"
                        - id: ContentEncAESSettings
                          header_size: 3
                          size: 7
                          children:
                            - id: AESSettingsCipherMode
                              header_size: 3
                              size: 4
                              value: AES-CTR
            - id: Video
              header_size: 2
              size: 18
              children:
                - id: PixelWidth
                  header_size: 2
                  size: 4
                  value: 360
                - id: PixelHeight
                  header_size: 2
                  size: 3
                  value: 240
                - id: DisplayWidth
                  header_size: 3
                  size: 5
                  value: 360
                - id: DisplayHeight
                  header_size: 3
                  size: 4
                  value: 240

